### PR TITLE
Add GLFW support to Vulkan window system integration

### PR DESCRIPTION
**Pull request by chances**
[GLFW Vulkan guide](https://www.glfw.org/docs/3.3/vulkan_guide.html)

- [x] Add `createVulkanGlfwSurface` to wsi module
- [x] Inegrate [`glfwGetPhysicalDevicePresentationSupport`](https://www.glfw.org/docs/3.3/vulkan_guide.html#vulkan_ext) into `VulkanPhysicalDevice.supportsSurface`
- [x] Add documentation of gfx's Vulkan Window System Integration support

### CODE REVIEW

File: vulkan/dub.json [hunk 1]
```diff
@@ -25,6 +25,13 @@
         {
             "name": "windows",
             "platforms": ["windows"]
+        },
+        {
+            "name": "glfw",
+            "platforms": ["linux", "windows"],
```
Comment (rtbo) [inline vulkan/dub.json:31]: don't forget `"versions": ["glfw"]`

**Comment by rtbo** [comment]
I'm not sure we need to alter the `gfx:window` package. glfw is doing a better job by itself.
If you can use gfx-d seamlessly out of glfw directly, this a big win.

**Comment by chances** [comment]
> I'm not sure we need to alter the `gfx:window` package. glfw is doing a better job by itself.

Cool; my gut was telling me just to leave it at the integration with the Vulkan sub-package of gfx, especially because my engine already has a working interface with GLFW itself. I'll revert my changes to the `gfx:window` package.

File: vulkan/source/gfx/vulkan/package.d [hunk 2]
```diff
@@ -147,6 +147,10 @@ VulkanInstance createVulkanInstance(VulkanCreateInfo createInfo=VulkanCreateInfo
     import std.range : chain;
     import std.string : toStringz;
 
+    // Surface instance extensions for the configured surface integration are mandatory so that
+    // the application may create surfaces
+    createInfo.mandatoryExtensions ~= surfaceInstanceExtensions;
```
Comment (chances) [inline vulkan/source/gfx/vulkan/package.d:152]: When configured with glfw, `surfaceInstanceExtensions` cannot be evaluated at compile-time. Also, surface extensions are mandatory for applications to function

File: vulkan/source/gfx/bindings/vulkan/vk.d [hunk 3]
```diff
@@ -6,13 +6,21 @@ version(Windows) {
     import core.sys.windows.windef : HINSTANCE, HWND;
 }
 
-version(linux) {
-    import xcb.xcb : xcb_connection_t, xcb_visualid_t, xcb_window_t;
+version(glfw) {
+    enum isGlfw = true;
+} else {
+    enum isGlfw = false;
 }
```
Comment (rtbo) [inline vulkan/source/gfx/bindings/vulkan/vk.d:13]: `vk.d` is auto generated, so you should not modify it in any way.
Current short term solution is to also have wayland and xcb dependencies (even if not used) when using glfw.

If you are interested, `vk.d` is generated by [vkdgen](https://github.com/rtbo/vkdgen)
If you need to remove the dependencies, I can introduce new versions like `HasWayland` and `HasXcb` to opt them in or out at compile time.

File: vulkan/source/gfx/vulkan/wsi.d [hunk 4]
```diff
@@ -36,77 +54,77 @@ enum swapChainDeviceExtension = "VK_KHR_swapchain";
...
-                inst.vk.CreateXcbSurfaceKHR(inst.vkObj, &sci, null, &vkSurf),
-                "Could not create Vulkan Xcb Surface"
-            );
-
-            return new VulkanSurface(vkSurf, inst);
+} else {
+    static if (!isGlfw) {
+        version(linux) {
+            import wayland.client : WlDisplay, WlSurface;
+            import xcb.xcb : xcb_connection_t, xcb_window_t;
```
Comment (rtbo) [inline vulkan/source/gfx/vulkan/wsi.d:61]: Same thing here, we can have versions like `HasGlfw`, `HasWayland`, `HasXcb`. (or `GfxGlfw`, `GfxWayland` and `GfxXcb` if we need namespace).
That way, the 3 possibilities can remain enabled at runtime depending on what is installed on target machine.

File: vulkan/source/gfx/vulkan/package.d (hunk 2 above)
Comment (rtbo) [reply vulkan/source/gfx/vulkan/package.d]: You're right, it is better at runtime anyway.
That way we can determine whether Wayland is available on the system

**Review by rtbo — APPROVED** [review]

**Comment by rtbo** [comment]
Thank you!
I've added some changes to have cleaner calculation of surface extensions at runtime.
Also I now check if Wayland is available.

**Pull request merged**